Contribute Guide on Deep Learning for Gesture Recognition on ESP32-S3

BlakeHansen130 opened the pull request:
## Description

Hello Espressif team,

This PR adds an article discussing the implementation of deep learning-based gesture recognition on the ESP32-S3, covering the journey from training to deployment. The goal of this article is to provide developers with a complete guide, helping them understand:
- How to train a deep learning model for gesture recognition.
- How to optimize the model for embedded deployment on ESP32-S3.
- How to deploy and run the model on the ESP32-S3.

I believe this article could serve as a valuable resource for developers looking to apply machine learning to embedded systems, and I'm eager to contribute to the Espressif community. 

## Related

This contribution is related to the discussion in issue #344. Please feel free to refer to it for more context.

## Testing

The article has undergone thorough grammar and content reviews to ensure clarity and accuracy. I have also tested the example code provided in the guide and verified that the steps work properly in a typical setup.

## Checklist

- [x] 🚨 This PR does not introduce breaking changes.
- [x] All CI checks (GH Actions) pass.
- [x] Documentation is updated as needed.
- [x] Tests are updated or added as necessary.
- [x] Code is well-commented, especially in complex areas.
- [x] Git history is clean — commits are squashed to the minimum necessary.

Please let me know if there are any areas where further improvements are needed or if there is additional information I can provide to make this article more helpful. I am open to any feedback and happy to make modifications as required.

Thank you very much for your time and consideration.

Best regards,  
BlakeHansen130

pedrominatel commented:
Thank you very much @BlakeHansen130 for your contribution!
We will start the review process and as soon as it's approved, this article will be published.

BlakeHansen130 commented:
Thank you very much for starting the review process! I'll keep an eye out for any feedback and will be happy to make any changes necessary.

BlakeHansen130 commented:
Thanks for the suggested fixes! I've updated the guide according to the review comments. Ready to merge after your approval.

BlakeHansen130 commented:
Sorry @pedrominatel, I accidentally clicked the refresh button on your review. I've already updated the date as well.

horw commented:
Hello, I think it would be great if you could attach a short demo video/images to show how it works/how it looks

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
````diff
@@ -0,0 +1,384 @@
+```python
+first_layer = Conv2d(1, 16, 3, stride=2)
+inverted_residual_blocks = [
+    (16, 24, stride=2, expand_ratio=6),
````
These are incorrect lines of code; maybe the class name was forgotten?

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
```diff
@@ -0,0 +1,384 @@
+
+*The content is organized into five main sections. The first section, "System Design," outlines the overall system architecture and development environment. The second section, "Model Development," addresses the design of the network architecture and the training strategy. The third section discusses techniques for "Quantization Optimization." The fourth section focuses on "Resource-Constrained D ...
+
+## System Design
```
I think this title is not very appropriate for the content inside it because it describes development environment and memory management(not that related to system design)...

Also, I found that the structure of this article jumps a lot between model development and model deployment.

As I understand, there are two parts to the development process for deep learning models on the ESP32:

  - On your computer: develop the model, train it, and validate it.
  - On your ESP32 device: upload the model and verify it.

It would be easier to read if this were more structured. 

Overall, the article is very nice. Thank you!

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
```diff
@@ -0,0 +1,384 @@
+    └── Channel for console output
+        └── Default: UART0
+
+Component config
```
I guess you mean this, because it's not related to Bluetooth.
```
(Top) → Component config → ESP System Settings
```

horw replied:
Could you also please clarify how it is related to this article?

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
```diff
@@ -0,0 +1,384 @@
+
+This configuration ensures proper compilation of the application components and correct linking with the ESP-DL framework. The integration of model files and headers follows a systematic approach through the build system.
+
+### Runtime Memory Management
```
Also, this content is about how to monitor memory, not about how it is managed. Could you clarify what is special about memory management here?

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
```diff
@@ -0,0 +1,384 @@
+
+This evaluation framework examines three critical aspects of each quantized model: memory footprint, accuracy retention, and inference latency. The modular design allows for easy comparison between different quantization strategies and helps in selecting the most appropriate approach for specific deployment requirements.
+
+Each quantization method presents its own set of trade-offs, and the choice between them depends on specific application requirements such as memory constraints, accuracy requirements, and inference speed needs. The systematic implementation and evaluation of these strategies ensure optimal deployment on the ESP32-S3 platform.
```
I guess you have some graph with information about the results of these quantized methods. It would be great if you could attach it here.

horw commented on content/blog/2024/12/Deep-Learning-for-Gesture-Recognition-on-ESP32-S3-from-Training-to-Deployment/index.md:
```diff
@@ -0,0 +1,384 @@
+
+## Experimental Results
+
+The gesture recognition system on ESP32-S3 showed strong performance across multiple metrics. Both quantitative and qualitative evaluations were conducted.
```
If 'multiple metrics' were mentioned here, they should also be specified somewhere in the text, along with what can be considered as strong.

BlakeHansen130 commented:
Thank you for your review and valuable comments, @horw. I have addressed your feedback with the following updates:

1. **Updated Model Definition**: I added the detailed modified code in the model definition section to provide more clarity.  
2. **Adjusted Subheadings**: Some subheading titles have been revised to make them more appropriate and consistent with the content.  
3. **Clarification on UART**: I explained the relevance of UART to the article, as it is related to the console output for debugging and configuring the ESP32-S3 during deployment.  
4. **Additional Visuals**: I included comparison charts of quantization results, images showing the model’s performance on the ESP32-S3, and a comprehensive workflow diagram.  

Please let me know if there’s anything further I can clarify or improve. Thank you again for your insightful feedback!

horw commented:
@BlakeHansen130 Thank you for the update. Overall, it is better than the previous version. However,  for your further contributions, I would suggest the following improvements in my opinion:

- The title and content should be more related.
- The structure should be easier to follow.
- The content should address the problem. If it's not relevant, perhaps you don’t need to mention it.
- When you mention something, you need to explain it, even if briefly.
- Avoid using general terms like "strong," "best," or "significantly." Provide concrete numbers; otherwise, the information is not valuable.
- Include code snippets only if they are very important. For example, if you're modeling a structure, maybe it's better to use an image.
- When you mention parameters, such as for deep learning layers or hyperparameters, try to explain why these numbers work better than others.

Consider whether your article offers enough information to effectively help readers solve the problem. Consider how your code snippets and images can help. When your article contains enough information to solve the related problem and helps me fully understand the steps you took and why, it will be a very valuable article.

P.S. In the ML field, I like how Andrej Karpathy explains things. For example, his explanation of transformer models: https://www.youtube.com/watch?v=kCc8FmEb1nY

f-hollow commented:
The folder name (URL slug) looks long, suggest shortening it from `deep-learning-for-gesture-recognition-on-esp32-s3-from-training-to-deployment/` to `esp32-s3-deep-learning-gesture-recongition`.

f-hollow commented:
Hi @BlakeHansen130 ,

I am Kirill - a technical writer at Espressif. I do not possess sufficient technical knowledge in your particular area, but I have some experience structuring knowledge, shaping content, and presenting it to readers, both in technical documentation and articles for developers.

I have read through your article and checked your project's [repo](https://github.com/BlakeHansen130/esp32s3-gesture-dl). From what I understand, your article describes the methodology that you used to implement your project and gives some insights into the steps you took and techniques you employed.

However, the purpose of your article is still not very clear. Let me explain what I mean.

You state clearly the following:

> This article provides an overview of the complete development process for a gesture recognition system, encompassing dataset preparation, model training, and deployment.

I see that you provide an overview of the development process. However, you cover only certain sides of the development process and other sides are only briefly touched on with phrases, such as `carefully designed to achieve optimal performance`, ` implemenmts a sophisticated memory management strategy`, and `strategically allocated to optimize performance`.

Usually before you write content, you decide on your target audience and the goal you want to achieve. I can see that your article might have the following potential purposes:

1. Science
    - **Target audience**: Scientific researchers looking for insights into optimizing the algorithms to enable running AI applications on Espressif SoCs.
    - **Your goal**: Share your achievements and benchmarks in optimizing algorithms and data processing.
2. Teaching
    - **Target audience**: Embedded developers who are interested in implementing AI applications on Espressif SoCs.
    - **Your goal**: Provide an easy-to-follow (ideally, as much as possible) tutorial that a regular developer can follow and achieve the same results as you have achieved.
3. Advertizing
    - **Target audience**: Potential buyers or investors for your product.
    - **Your goal**: Demonstrate what you achieved and how it can be used commercially in order to earn income or attract investments.

Let's quickly go through all the purposes and how they compare against your current state of the article:

1. **Science**: I am not very experienced with scientific and academic writing, but I can assume that you need to focus on certain achievements, present the current state-of-the-art benchmarks from reputable sources and share your benchmarks. In this way, it will be clear that you have proven methods to achieve better results that might be of interest to this target audience.
2. **Teaching**: You only provide broad statements and a repo with data and code. A regular developer would have no idea what to do to teach their own model and deploy it on an Espressif SoC. If this is what you want, you need to turn your article into a tutorial - a number of reproducible steps that a developer can follow to achieve the goal of having a binary on your ESP32 that does gesture recognition without requiring too much research on the part of the developer (that's exactly what tutorials are for).
3. **Advertizing**: If you want to advertize your advances, first and foremost, you need to provide some evidence that demonstrates your claims: the ESP32 recognizing your hand gestures in real time. For that, you might need to provide a video demonstration and maybe a way to download your binary and flash it onto a chip, so that others can try it and confirm that it works.

Now after this analysis let's get back to your article. Can you please tell me what target audience and what your goal is?

- If you don't want to look at it this way, please share your vision on what you want to achieve with your content.
- If you simply didn't think about the purpose, it is never too late to decide on what you want and restucture your article accordingly. I will gladly help you with that as an editor.

Now please let me know what you think.

BlakeHansen130 commented:
@f-hollow Thank you for the feedback! During my undergraduate design project, I was inspired by Bukharai's gesture recognition [tutorial](https://developer.espressif.com/blog/hand-gesture-recognition-on-esp32-s3-with-esp-deep-learning/). After successfully implementing the complete workflow with newer versions of esp-idf, esp-dl and esp-ppq (as discussed in https://github.com/alibukharai/Blogs/issues/10), I wanted to share this article to help others exploring similar implementations.

BlakeHansen130 commented:
@horw Thank you for your feedback. Here are my implemented changes:

1. Updated the title to better match content
2. Added detailed model structure images
3. Replaced subjective terms with concrete metrics

Regarding the structure - I've maintained it as is because it directly mirrors my project's implementation workflow. Each section progresses naturally from data preprocessing through model development and quantization to deployment, just like the organized hierarchy in my code repository. This parallel between article structure and actual development process helps readers understand both the theoretical concepts and practical implementation steps simultaneously.

I'm also a fan of 3Blue1Brown's excellent teaching style and continue learning from his videos, and I'm still working toward that level of clarity in my explanations.

f-hollow commented:
@BlakeHansen130 Hi, it's been a while.

We recently merged the article [Touchpad Digit Recognition Based on ESP-DL](https://developer.espressif.com/blog/2025/06/touchpad-digit-recognition/), and it is a good example of how such articles are usually written.

Do you think you will have a chance to make your article more insightful and helpful?

Please let us know. If not, we will be OK merging your article as it is.

f-hollow commented:
@BlakeHansen130 

I will close this PR for now. Feel free to reply to my previous message, and we will reopen this PR.
